<p style="font-family: 'Meiryo UI'; font-size: 32px; padding: 25px; text-align: center; color: #ffffff; border-radius: 25px;  font-weight: bold; background-color: #005080;">📚 Hyperparameter Optimization with the MS-EPSO 🐦</p>

## Hyperparameter Optimization Problem

Hyperparameter optimization is the process of finding the optimal set of hyperparameters that result in the best performance of a model. This is an important step in the machine learning pipeline as it can lead to significant improvements in model accuracy and generalization.

Traditional methods for hyperparameter optimization include:

1. **Grid Search**: This method involves exhaustively searching through a manually specified subset of the hyperparameter space. While simple to implement, grid search is computationally expensive and not feasible for large hyperparameter spaces.

2. **Bayesian Optimization**: This method builds a probabilistic model of the objective function and uses it to select the most promising hyperparameters to evaluate in the true objective function. While more efficient, it can be complex to implement and computationally intensive.

In this project, we will test my published Swarm Intelligence algorithm named [Maximum Search Limitations Evolutionary Particle Swarm Optimization (MS-EPSO)](https://link.springer.com/chapter/10.1007/978-3-030-30241-2_59) for hyperparameter tuning.

<p style="font-family: 'Meiryo UI'; font-size: 30px; padding: 12px; text-align: center; color: #ffffff; border-radius: 15px;  font-weight: bold; background-color: #007040;">📚 Libraries</p>

In this experiment, we will use the XGBoost algorithm as our base model. We will utilize the original implementation of the MS-EPSO algorithm, which can be found [here](https://github.com/MtrsN/MS-EPSO).

In [1]:
import json

import numpy as np
import pandas as pd
import xgboost as xgb

from msepso import MSEPSO
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

<p style="font-family: 'Meiryo UI'; font-size: 30px; padding: 12px; text-align: center; color: #ffffff; border-radius: 15px;  font-weight: bold; background-color: #007040;">📊 Dataset</p>

In this notebook, we will use the Breast Cancer dataset. This dataset is simple and comes pre-processed. We will load the data from the sklearn environment.

In [2]:
data = load_breast_cancer()

df = pd.DataFrame(data.data, columns=data.feature_names)

df['Target'] = data.target

df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  Target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [3]:
print("The dataset has {} rows and {} columns".format(df.shape[0], df.shape[1]))

The dataset has 569 rows and 31 columns


<p style="font-family: 'Meiryo UI'; font-size: 30px; padding: 12px; text-align: center; color: #ffffff; border-radius: 15px;  font-weight: bold; background-color: #007040;">Why MS-EPSO❓</p>

In recent years, Optuna has established itself as a valuable hyperparameter optimization framework. This framework allows the implementation of various state-of-the-art optimization methods for tuning machine learning models. Among these techniques, the CMA-ES algorithm is notably included.

MS-EPSO was developed to address specific issues found in the Evolutionary Particle Swarm Optimization (EPSO) algorithm. MS-EPSO enhances the balance between exploration (searching through the parameter space) and exploitation (refining the best solutions found). 

MS-EPSO was initially tailor made for a specific application, but it has proven to be a viable solution for general optimization problems, particularly those where the variables exhibit low standard deviation. For instance, it is highly effective in Feature Selection problems, where the decision is binary (i.e., to include or not include a variable). This makes MS-EPSO an interesting choice for tasks.

<p style="font-family: 'Meiryo UI'; font-size: 30px; padding: 12px; text-align: center; color: #ffffff; border-radius: 15px;  font-weight: bold; background-color: #007040;">🚩 The Problem</p>

In order to apply MS-EPSO to any problem, we have to define:

1. Objective function
2. Boundaries
3. Algorithm parameters

However, the hyperparameter problem requires several settings, not limited to:

1. Model
2. Model Parameters
3. Metric
4. Dataset

For a better understanding of each parameter, check the XGBoost documentation and the MS-EPSO algorithm class implemented in the `msepso.py` file

In [4]:
'''Split the data into training, validation, and test sets. With a larger dataset, we could check for the possibility to stratify the split based on the target variable. 
This would ensure that the distribution of the target variable is similar across all the splits. However, in this case, we will keep it simple and split the data randomly.
'''

RANDOM_SEED = 101

X = df.drop('Target', axis=1)
y = df['Target']

X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size= 0.33, random_state=RANDOM_SEED)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size= 0.5, random_state=RANDOM_SEED)

print(f'Training set: {X_train.shape}, {y_train.shape}')
print(f'Validation set: {X_val.shape}, {y_val.shape}')
print(f'Test set: {X_test.shape}, {y_test.shape}')

Training set: (381, 30), (381,)
Validation set: (94, 30), (94,)
Test set: (94, 30), (94,)


In [5]:
# Create our objective function based on MS-EPSO parameters and XGBoost performance.

def objective_function(x):

    policy_index = int(x[9])

    params = {

        'verbosity': 0,
        'seed' : RANDOM_SEED,
        'eval_metric': 'logloss',

        'max_depth': int(x[0]),
        'min_child_weight': int(x[1]),
        'n_estimators': int(x[2]),
        'gamma': x[3],
        'subsample': x[4],
        'reg_alpha': x[5],
        'reg_lambda': x[6],
        'learning_rate': x[7],
        'colsample_bytree': x[8],
        'grow_policy':  ["depthwise", "lossguide"][policy_index]
    }

    model = xgb.XGBClassifier(**params)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_val)
    
    acc = accuracy_score(y_val, y_pred)

    return acc

# Create our problem boundaries
lower_bounds = np.array([0, 1, 50, 1e-9, 0.3, 1e-9, 1e-9, 0.01, 0.3, 0])
upper_bounds = np.array([12, 7, 100, 0.5, 1.0, 1.0, 1.0, 1.0, 1.0, 1])

# this is just one example for you to understand what MS-EPSO will try to find =) (Not exactly those numbers)
example_solution = [
        0.5, # max_depth
        3, # min_child_weight
        1, # n_estimators
        0.5, # gamma
        0.5, # subsample
        0.5, # reg_alpha
        0.5, # reg_lambda
        0.5, # learning_rate
        0.5, # colsample_bytree
        1 # grow_policy
]

In [6]:
%%time

# MS-EPSO parameters: We reduce the number of maximum function evaluationspeed up the optimization process.

ms_parameters = {
    "tau" : 0.8,
    "cp" : 0.9,
    "nr" : 1,
    "mll" : 25,
    "max_fes" : 1000,
    "num_sol" : 100,
    "init_method" : "uniform",
    "seed" : 101
}

msepso = MSEPSO(fx= objective_function, ub= upper_bounds, lb= lower_bounds, **ms_parameters)
msepso.run()

print("\n\n")

print("Best solution: ", msepso.best_particle)
print("Best fitness: ", msepso.best_particle_fitness)

print("\n\n")

MS-EPSO - Function Evaluations: 100%|█████████▉| 898/900 [00:52<00:00, 23.65it/s]




Best solution:  [ 9.59593639  6.25972995 65.13119009  0.30463583  0.35950076  0.31917783
  0.81549922  0.87368236  0.92349529  0.6899099 ]
Best fitness:  0.8829787234042553



CPU times: total: 9min 45s
Wall time: 52.2 s


In [7]:
best_parameters = {

    'verbosity': 0,
    'seed' : RANDOM_SEED,
    'eval_metric': 'logloss',

    'max_depth': int(msepso.best_particle[0]),
    'min_child_weight': int(msepso.best_particle[1]),
    'n_estimators': int(msepso.best_particle[2]),
    'gamma': msepso.best_particle[3],
    'subsample': msepso.best_particle[4],
    'reg_alpha': msepso.best_particle[5],
    'reg_lambda': msepso.best_particle[6],
    'learning_rate': msepso.best_particle[7],
    'colsample_bytree': msepso.best_particle[8],
    'grow_policy':  ["depthwise", "lossguide"][int(msepso.best_particle[9])]
}

model = xgb.XGBClassifier(**best_parameters)
model.fit(X_train, y_train)

preds = model.predict(X_test)

acc = accuracy_score(y_test, preds)

print(f"Accuracy on test set: {acc}")

Accuracy on test set: 0.9042553191489362


<p style="font-family: 'Meiryo UI'; font-size: 30px; padding: 12px; text-align: center; color: #ffffff; border-radius: 15px;  font-weight: bold; background-color: #007040;">📈 Results</p>

In our experiment, we utilized the MS-EPSO algorithm to tune the hyperparameters of the XGBoost model using the Breast Cancer dataset.

The model was trained with these parameters and evaluated on the test set. The accuracy achieved on the test set was 90.43%. This result is not outstanding for the current dataset, however, this is a short experiment without many modifications to existing code. It at minimum points out how MS-EPSO can be applied to optimizing hyperparameters and can potentially lead to great results.

In [8]:
print(json.dumps(best_parameters, indent=4), end= "\n\n")
print(f"Accuracy on test set: {acc}")

{
    "verbosity": 0,
    "seed": 101,
    "eval_metric": "logloss",
    "max_depth": 9,
    "min_child_weight": 6,
    "n_estimators": 65,
    "gamma": 0.30463583292459473,
    "subsample": 0.3595007585087136,
    "reg_alpha": 0.319177826731411,
    "reg_lambda": 0.8154992172626163,
    "learning_rate": 0.8736823566749992,
    "colsample_bytree": 0.9234952864984858,
    "grow_policy": "depthwise"
}

Accuracy on test set: 0.9042553191489362


<p style="font-family: 'Meiryo UI'; font-size: 30px; padding: 12px; text-align: center; color: #ffffff; border-radius: 15px;  font-weight: bold; background-color: #007040;">🙂 Conclusion</p>

This experiment aimed to demonstrate the application of the MS-EPSO algorithm for hyperparameter tuning in the XGBoost model using the Breast Cancer dataset. The MS-EPSO algorithm successfully optimized several hyperparameters, resulting in a model that achieved an accuracy of 90.43% on the test set.

Although the results were not the best possible, the experiment was brief and conducted within a limited scope. This highlights the potential of MS-EPSO, even in a constrained setting.

There is ongoing debate about the best applications for algorithms of this category across different problems. From my perspective, Swarm Intelligence algorithms, like MS-EPSO, are particularly useful when a better solution is not readily available. For example, in this case, while state-of-the-art frameworks like Optuna (which I am currently using in a Kaggle competition) are excellent choices for hyperparameter optimization, Swarm Intelligence algorithms offer a robust alternative when conventional methods fall short. According to the No Free Lunch theorem, no single algorithm can solve all problems optimally. Therefore, when you are unsure how to approach a problem, exploring solutions through Swarm Intelligence or evolutionary computation algorithms can be highly beneficial.

Thank you for reading this notebook. I appreciate your time and hope you learned something new. See you soon! 😊